### ISSUE TITLE
LibreOffice flatpak doesn't stop "saving" where it can't write to, then "deletes" the "file" that never existed

### ISSUE THREAD

**Issue by Githob2021** [flathub/org.libreoffice.LibreOffice#176]
Description:
Currently LibreOffice flatpak does not provide an error/warning prompt when it is not able to save to a location it does not have access to view flatseal/bubblewrap.

What occurs, is the user is able to use File menu Save as and pick any location (even if LibreOffice flatpak does not have access) to save in the file picker window that appears. Even if you use that file picker to do another save as or open you can see the file that was created along with the .lock file.

However, if the user looks with their own integrated file explorer which is Dolphin for me they will see the files were never actually created. Neither the regular file or .lock.

This makes me curious if this is a bug in:
1) LibreOffice Flatpak
2) Integrated file picker/explorer
3) Bubblewrap/Flatseal
4) Flatpak itself

It is concerning that LibreOffice flatpak "creates" a file that can be seen in the file picker/explorer pop up but not outside of it.

Steps to Reproduce:
1. Create a document in LibreOffice Calc flatpak
2. Write some stuff
3. Click File and Save as in the menu bar. "save" it to a location there is no access via flatseal LibreOffice Calc "thinks" the file is saved there as the file picker suggets
4. Navigate to that location in your file explorer (Dolphin for me).
5. Close LibreOffice Calc flatpak
6. Reopen LibreOffice Calc flatpak
7. The file never existed in LibreOffice Calc apparently
8. Cry because you lost hours of work.

Actual Results:
The file is nowhere to be found.

Expected Results:
LibreOffice would warn the user that they are not able to save to that directory due to lacking permissions in flatseal/bubblewrap.

Reproducible: Always
User Profile Reset: Yes
OpenGL enabled: Yes

Additional Info:
Operating System: Arch Linux
KDE Plasma Version: 5.23.5
KDE Frameworks Version: 5.90.0
Qt Version: 5.15.2
Kernel Version: 5.15.16-hardened1-1-hardened (64-bit)
Graphics Platform: X11

**Comment by Erick555** [comment]
> Click File and Save as in the menu bar. "save" it to a location there is no access via flatseal LibreOffice Calc "thinks" the file is saved there as the file picker suggets

I'm unable to reproduce this part. LO doesn't use portal so in the file picker it shows only directiories it can write into (unless you made something read-only via flatseal). Could you post` flatpak info -M org.libreoffice.LibreOffice` output?

**Comment by Githob2021** [comment]
> > Click File and Save as in the menu bar. "save" it to a location there is no access via flatseal LibreOffice Calc "thinks" the file is saved there as the file picker suggets
> 
> I'm unable to reproduce this part. LO doesn't use portal so in the file picker it shows only directiories it can write into (unless you made something read-only via flatseal). Could you post` flatpak info -M org.libreoffice.LibreOffice` output?

```
flatpak info -M org.libreoffice.LibreOffice
[Context]
shared=ipc;
sockets=x11;wayland;
devices=dri;
filesystems=/home/USERNAME/Downloads;!host;

[Session Bus Policy]
com.canonical.AppMenu.Registrar=talk
org.gtk.vfs.*=talk
org.libreoffice.LibreOfficeIpc0=own

[Environment]
GIO_EXTRA_MODULES=/app/lib/gio/modules
GTK_THEME=Adwaita:dark
JAVA_HOME=/app/jre
LIBO_FLATPAK=1
```

> Where exactly you are saving those files? If it's /tmp then it makes sense as flatpak has separate /tmp from host which is discarded after app close.

I have been saving them to my Downloads folder.

**Comment by Erick555** [comment]
Then I can' reproduce this. When running `flatpak run --filesystem=/home/user/Downloads --nofilesystem=host org.libreoffice.LibreOffice` file picker shows only my Downloads folder and I can save files there successfully.

**Comment by Githob2021** [comment]
> Then I can' reproduce this. When running `flatpak run --filesystem=/home/user/Downloads --nofilesystem=host org.libreoffice.LibreOffice` file picker shows only my Downloads folder and I can save files there successfully.

Can you please attach a screenshot of your file picker? I wanted to compare it to mine/screenshots I provided above.

**Comment by Erick555** [comment]
> I have been saving them to my Downloads folder and clicked up one level to my Home folder where it didn't have access. The problem was LibreOffice did not warn me it didn't actually save but I can see it in the file picker.

If you mean that you saved the file to `/home/<something.ods>` then it should be successfully saved (you should be able to open the file if you didn't close the app) but after app is closed then the file will be gone because the parent /home directory isn't from host. There's no bug here.

**Comment by Erick555** [comment]
Yeah so your screenshots show what I descried above - those files were actually saved but this directory was a temporary filesystem existing only in flatpak sandbox and closing app discarded it.

**Comment by Githob2021** [comment]
> Yeah so your screenshots show what I described above - those files were actually saved but this directory was a temporary filesystem existing only in flatpak sandbox and closing app discarded it.
> 
> You may try saving the document then closing it (without closing whole app) then you should be able to re-open it until you close the app as a whole.

It may be a niche use case, but don't you think LibreOffice (flatpak only) should make a backup somewhere in case of this? Accidents do happen, like it did for me. If not, feel free to close this ticket. Thank you either way.

**Comment by Erick555** [comment]
Sometimes users may want to saves files in temporary filesystems (privacy, etc.) and if LO backed up those somewhere in hard drive then those users won't be pleased :smile: . Personally I don't see what LO could do if it doesn't understand what the user wanted in the end. Saving file in /tmp from non-flatpak LO then rebooting would have same effect (lost files).

**Comment by Githob2021** [comment]
> Personally I don't see what LO could do if it doesn't understand what the user wanted in the end

Why not have a setting that would expose this where a user could choose if they want to have a back up created. Those who want privacy could turn it off, those that want redundancy can turn it on.

**Comment by Erick555** [comment]
Regardless if it's technically possible I bet you would learn about this setting after you lost your files not before :smile:

**Comment by WineBottles** [comment]
> Regardless if it's technically possible I bet users would learn about this setting after they lost their files not before smile

This is kind of what I meant about a safety net:
https://twitter.com/usebottles/status/1489307268235636737?cxt=HHwWgsC9ybvEiqspAAAA
In Bottles, it will warn you if it cannot access the directory.

"Who asked for custom bottles path? 👀 [#Linux](https://twitter.com/hashtag/Linux?src=hashtag_click) 
Oh and don't worry about Flatpak, if you forget to give permissions it will warn you and fallback to default"

What I was trying to say earlier was it should have a fallback aka backup location where it saves if it cannot actually save to the location because it's "temp" as you mentioned.

**Comment by takluyver** [comment]
This is probably only solvable by Libreoffice using the FileChooser portal to open/save files.

Flatpak deliberately sets up a temporary filesystem to run applications in, so that a compromised application (e.g. if you open a malicious document that exploits a vulnerability in LibreOffice) can't read & write arbitrary files on your system. But when the user goes to save or open a file, applications can use a 'portal' to show users their real filesystem and let them choose a file to use.

TL;DR: making a sandbox for applications that are used to running without one is hard. :disappointed:

**Comment by Erick555** [comment]
This issue is about saving files in paths not covefred by `--filesystem=host` so your first comment was right - the solution is to use FileChooser portal by LO.

**Comment by nekohayo** [comment]
> This is probably only solvable by Libreoffice using the FileChooser portal to open/save files.

As I understand it, this would be:
* https://bugs.documentfoundation.org/show_bug.cgi?id=159311
* https://bugs.documentfoundation.org/show_bug.cgi?id=165028
